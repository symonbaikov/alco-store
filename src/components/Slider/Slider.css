.slider {
  position: relative;
  height: 600px;
  overflow: hidden;
  margin: 0;
  padding: 0;
  display: flex;
  justify-content: center;
  pointer-events: none;
}

.slider-container {
  margin-top: 30px;
  position: relative;
  width: 1500px;
  height: 100%;
  margin-left: auto;
  margin-right: auto;
  pointer-events: auto;
}

.slide {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  transition: opacity 0.5s ease-in-out;
  background-size: 100% 100%;
  background-repeat: no-repeat;
  background-position: center;
  display: flex;
  align-items: center;
  padding-left: 100px;
  background-color: #f5f5f5;
}

.slide.active {
  opacity: 1;
  z-index: 1;
}

.slide-content {
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  padding: 40px;
  background: linear-gradient(to top, rgba(0, 0, 0, 0.7), transparent);
  color: white;
}

.slide-content h2 {
  font-size: 32px;
  margin-bottom: 10px;
  font-weight: 600;
  font-family: "Poppins", sans-serif;
}

.slide-content p {
  font-size: 18px;
  margin-bottom: 20px;
}

.slide-button {
  display: inline-block;
  padding: 12px 30px;
  background: #8b0000;
  color: white;
  font-family: Poppins, sans-serif;
  text-decoration: none;
  font-weight: 600;
}

.slider:hover .slider-button {
  opacity: 1;
  transform: translateY(-50%) translateX(0);
}

.slider-button {
  position: absolute;
  top: 50%;
  width: 40px;
  height: 50px;
  background: rgba(255, 255, 255, 0.9);
  border: none;

  color: #333;
  font-size: 20px;
  cursor: pointer;
  z-index: 10;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 0;
  transition: all 0.3s ease;
  box-shadow: 0 2px 6px rgba(0, 0, 0, 0.2);
  pointer-events: auto;
}

.slider-button:hover {
  background: white;
  color: #8b0000;
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.3);
}

.slider-button.prev {
  left: 120px;
  transform: translateY(-50%) translateX(-20px);
}

.slider-button.next {
  right: 120px;
  transform: translateY(-50%) translateX(20px);
}

.slider-dots {
  position: absolute;
  bottom: 20px;
  left: 50%;
  transform: translateX(-50%);
  display: flex;
  gap: 8px;
  z-index: 10;
  pointer-events: auto;
}

.dot {
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background: rgba(255, 255, 255, 0.5);
  border: none;
  padding: 0;
  cursor: pointer;
  transition: all 0.3s ease;
  pointer-events: auto;
}

.dot.active {
  background: white;
  transform: scale(1.2);
}

@media (max-width: 768px) {
  .slider {
    height: 400px;
  }

  .slider-container {
    width: 100%;
    margin-top: 0;
  }

  .slide {
    padding-left: 20px;
  }

  .slide-content {
    padding: 20px;
  }

  .slide-content h2 {
    font-size: 24px;
  }

  .slide-content p {
    font-size: 16px;
  }

  .slider-button {
    width: 40px;
    height: 40px;
    font-size: 16px;
  }

  .slider-button.prev {
    left: 20px;
  }

  .slider-button.next {
    right: 20px;
  }

  .slider-dots {
    bottom: 10px;
  }

  .dot {
    width: 6px;
    height: 6px;
  }
}
