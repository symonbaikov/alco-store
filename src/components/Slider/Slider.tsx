import { useState, useEffect } from "react";
import { useTranslation } from "react-i18next";
import { getSlides } from "../../../server/services/slides-service";
import SliderSkeleton from "./SliderSkeleton";
import "./Slider.css";

interface Slide {
  id: number;
  image: string;
  title: string;
  description: string;
  link: string;
  order: number;
}

export const Slider = () => {
  const { t } = useTranslation();
  const [slides, setSlides] = useState<Slide[]>([]);
  const [currentSlide, setCurrentSlide] = useState(0);
  const [isPaused, setIsPaused] = useState(false);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const loadSlides = async () => {
      try {
        const data = await getSlides();
        setSlides(data);
      } catch (error) {
        console.error("Error loading slides:", error);
      } finally {
        setIsLoading(false);
      }
    };
    loadSlides();
  }, []);

  useEffect(() => {
    if (!isPaused && slides.length > 0) {
      const timer = setInterval(() => {
        setCurrentSlide((prev) => (prev + 1) % slides.length);
      }, 5000);
      return () => clearInterval(timer);
    }
  }, [isPaused, slides.length]);

  const goToSlide = (index: number) => {
    setCurrentSlide(index);
  };

  const nextSlide = () => {
    setCurrentSlide((prev) => (prev + 1) % slides.length);
  };

  const prevSlide = () => {
    setCurrentSlide((prev) => (prev - 1 + slides.length) % slides.length);
  };

  if (isLoading) {
    return <SliderSkeleton />;
  }

  if (slides.length === 0) {
    return null;
  }

  return (
    <div
      className="slider"
      role="region"
      aria-label="slider"
      onMouseEnter={() => setIsPaused(true)}
      onMouseLeave={() => setIsPaused(false)}
    >
      <div className="slider-container">
        {slides.map((slide, index) => (
          <div
            key={slide.id}
            className={`slide ${index === currentSlide ? "active" : ""}`}
            style={{ backgroundImage: `url(${slide.image})` }}
            role="article"
            aria-label={`slide ${index + 1}`}
            aria-hidden={index !== currentSlide}
          >
            <div className="slide-content">
              <h2>{t(`slides.${slide.order}.title`)}</h2>
              <p>{t(`slides.${slide.order}.description`)}</p>
              <a href={slide.link} className="slide-button">
                {t('common.moreInfo')}
              </a>
            </div>
          </div>
        ))}
      </div>

      <button 
        className="slider-button prev" 
        onClick={prevSlide}
        aria-label="previous slide"
      >
        <i className="fas fa-chevron-left"></i>
      </button>
      <button 
        className="slider-button next" 
        onClick={nextSlide}
        aria-label="next slide"
      >
        <i className="fas fa-chevron-right"></i>
      </button>

      <div className="slider-dots" role="tablist">
        {slides.map((_, index) => (
          <button
            key={index}
            className={`dot ${index === currentSlide ? "active" : ""}`}
            onClick={() => goToSlide(index)}
            role="tab"
            aria-selected={index === currentSlide}
            aria-label={`Go to slide ${index + 1}`}
          />
        ))}
      </div>
    </div>
  );
};