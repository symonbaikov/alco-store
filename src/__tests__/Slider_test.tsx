import { render, act, waitFor } from '@testing-library/react';
import { Slider } from '../components/Slider/Slider';

// Мокаем модуль сервиса слайдов
jest.mock('server/services/slides-service', () => ({
  getSlides: jest.fn(() => Promise.resolve([
    {
      id: 1,
      image: '/images/slide1.jpg',
      title: 'Slide Title 1',
      description: 'Slide Description 1',
      link: '/link1',
      order: 1
    },
    {
      id: 2,
      image: '/images/slide2.jpg',
      title: 'Slide Title 2',
      description: 'Slide Description 2',
      link: '/link2',
      order: 2
    }
  ]))
}));

describe('Slider Component', () => {
  // Снапшот-тест
  test('matches snapshot', async () => {
    const { container } = render(<Slider />);
    
    // Ждем загрузки слайдов
    await waitFor(() => {
      expect(container.querySelector('.MuiSkeleton-root')).not.toBeInTheDocument();
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  // Интеграционные тесты
  test('renders slider component and loads slides', async () => {
    const { container } = render(<Slider />);

    // Проверяем состояние загрузки
    expect(container.querySelector('.MuiSkeleton-root')).toBeInTheDocument();

    // Ждем загрузки слайдов
    await waitFor(() => {
      expect(container.querySelector('.MuiSkeleton-root')).not.toBeInTheDocument();
    });

    // Проверяем, что слайды загрузились
    const slides = container.querySelectorAll('.slide');
    expect(slides.length).toBe(2);

    // Проверяем навигационные кнопки
    const prevButton = container.querySelector('.slider-button.prev');
    const nextButton = container.querySelector('.slider-button.next');
    expect(prevButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();

    // Проверяем точки навигации
    const dots = container.querySelectorAll('.dot');
    expect(dots.length).toBe(2);
  });

  test('handles slide navigation', async () => {
    const { container } = render(<Slider />);
    
    // Ждем загрузки слайдов
    await waitFor(() => {
      expect(container.querySelector('.MuiSkeleton-root')).not.toBeInTheDocument();
    });

    const nextButton = container.querySelector('.slider-button.next') as HTMLButtonElement;
    expect(nextButton).not.toBeNull();
    
    // Проверяем начальное состояние
    const slides = container.querySelectorAll('.slide');
    expect(slides[0].classList.contains('active')).toBe(true);
    expect(slides[1].classList.contains('active')).toBe(false);
    
    // Кликаем на следующий слайд
    await act(async () => {
      nextButton.click();
    });

    // Проверяем, что второй слайд стал активным
    expect(slides[1].classList.contains('active')).toBe(true);
    expect(slides[0].classList.contains('active')).toBe(false);
  });
});