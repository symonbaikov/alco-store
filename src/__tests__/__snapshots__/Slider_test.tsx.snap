// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Component matches snapshot 1`] = `
<div
  aria-label="slider"
  class="slider"
  role="region"
>
  <div
    class="slider-container"
  >
    <div
      aria-hidden="false"
      aria-label="slide 1"
      class="slide active"
      role="article"
      style="background-image: url(/images/slide1.jpg);"
    >
      <div
        class="slide-content"
      >
        <h2>
          slides.1.title
        </h2>
        <p>
          slides.1.description
        </p>
        <a
          class="slide-button"
          href="/link1"
        >
          common.moreInfo
        </a>
      </div>
    </div>
    <div
      aria-hidden="true"
      aria-label="slide 2"
      class="slide "
      role="article"
      style="background-image: url(/images/slide2.jpg);"
    >
      <div
        class="slide-content"
      >
        <h2>
          slides.2.title
        </h2>
        <p>
          slides.2.description
        </p>
        <a
          class="slide-button"
          href="/link2"
        >
          common.moreInfo
        </a>
      </div>
    </div>
  </div>
  <button
    aria-label="previous slide"
    class="slider-button prev"
  >
    <i
      class="fas fa-chevron-left"
    />
  </button>
  <button
    aria-label="next slide"
    class="slider-button next"
  >
    <i
      class="fas fa-chevron-right"
    />
  </button>
  <div
    class="slider-dots"
    role="tablist"
  >
    <button
      aria-label="Go to slide 1"
      aria-selected="true"
      class="dot active"
      role="tab"
    />
    <button
      aria-label="Go to slide 2"
      aria-selected="false"
      class="dot "
      role="tab"
    />
  </div>
</div>
`;
